File: cpp/src/micro_functions_bench.cpp
/// Individual function micro benchmarks — prove each function is equally fast in C++ and Rust.
///
/// Usage: ./cpp_micro_functions [iterations]
/// Default: 10 iterations, 1,000,000 operations each

#include <cstdio>
#include <cstdint>
#include <cstring>
#include <cstdlib>
#include <chrono>
#include <vector>
#include <string>
#include <random>

#include "column_marshaller.h"

static constexpr size_t NUM_OPS = 1000000;
static constexpr size_t DEFAULT_ITERS = 10;

using Clock = std::chrono::high_resolution_clock;
static double elapsed_ms(Clock::time_point t0, Clock::time_point t1) {
    return std::chrono::duration<double, std::milli>(t1 - t0).count();
}

// ─── 1. arena_alloc_only ─────────────────────────────────────────────────
__attribute__((noinline))
static uint64_t RunArenaAllocOnly() {
    taper::SimpleArenaAllocator pool;
    uint64_t checksum = 0;
    for (size_t i = 0; i < NUM_OPS; i++) {
        uint8_t* p = pool.Allocate(50);
        checksum += reinterpret_cast<uint64_t>(p);
    }
    return checksum;
}

// ─── 2. new_row_only ─────────────────────────────────────────────────────
__attribute__((noinline))
static uint64_t RunNewRowOnly() {
    taper::SimpleArenaAllocator pool;
    std::vector<size_t> keySizes = {0, 0, 0, 0};
    std::vector<taper::ColumnKind> kinds(4, taper::ColumnKind::Varchar);
    taper::RowContainer rc(keySizes, kinds, 8, pool);
    uint64_t checksum = 0;
    for (size_t i = 0; i < NUM_OPS; i++) {
        char* row = rc.NewRow();
        checksum += reinterpret_cast<uint64_t>(row);
    }
    return checksum;
}

// ─── 3. serialize_4str_only ──────────────────────────────────────────────
__attribute__((noinline))
static uint64_t RunSerialize4strOnly(const std::vector<std::vector<taper::VarcharSlice>>& slices) {
    taper::SimpleArenaAllocator pool;
    uint64_t checksum = 0;
    for (size_t i = 0; i < NUM_OPS; i++) {
        size_t totalSize = 0;
        for (size_t c = 0; c < 4; c++) {
            totalSize += 1 + taper::ComputeRowLenSize(slices[c][i].len) + slices[c][i].len;
        }
        uint8_t* block = pool.Allocate(static_cast<int64_t>(totalSize));
        uint8_t* wp = block;
        for (size_t c = 0; c < 4; c++) {
            wp += taper::SerializeVarcharToBuffer(wp, slices[c][i].ptr, slices[c][i].len);
        }
        checksum += reinterpret_cast<uint64_t>(block);
    }
    return checksum;
}

// ─── 4. compare_4str_only ────────────────────────────────────────────────
__attribute__((noinline))
static uint64_t RunCompare4strOnly(
    const std::vector<const uint8_t*>& arenaBlocks,
    const std::vector<std::vector<taper::VarcharSlice>>& slices)
{
    uint64_t match_count = 0;
    for (size_t i = 0; i < NUM_OPS; i++) {
        const uint8_t* pos = arenaBlocks[i];
        bool all_match = true;
        for (size_t c = 0; c < 4; c++) {
            if (!taper::CompareVarcharFromRow(pos, slices[c][i].ptr, slices[c][i].len)) {
                all_match = false;
                break;
            }
            pos += taper::ComputeVarCharSerializedSize(pos);
        }
        if (all_match) match_count++;
    }
    return match_count;
}

// ─── 5. hashmap_probe_only ───────────────────────────────────────────────
__attribute__((noinline))
static uint64_t RunHashmapProbeOnly(const std::vector<int64_t>& hashes, size_t numChunks) {
    taper::TaperFlatHashTable table(numChunks);
    uint64_t sum = 0;
    table.EmplaceBatch(hashes.data(), static_cast<int32_t>(hashes.size()),
        [](int32_t) { return false; },
        [](uint32_t, char* data) { memset(data, 0x42, taper::ROW_PTR_SIZE); },
        [&sum](uint32_t, char*, bool isNew) { if (!isNew) sum++; }
    );
    return table.Size() + sum;
}

// ─── 6. setrowptr_only ──────────────────────────────────────────────────
__attribute__((noinline))
static uint64_t RunSetGetRowPtrOnly() {
    char buf[6];
    uint64_t checksum = 0;
    for (size_t i = 0; i < NUM_OPS; i++) {
        uint8_t* ptr = reinterpret_cast<uint8_t*>(0x7FFF00000000ULL + i * 41);
        taper::SetRowPtr(buf, ptr);
        uint8_t* got = taper::GetRowPtr(buf);
        checksum += reinterpret_cast<uint64_t>(got);
    }
    return checksum;
}

// ═══════════════════════════════════════════════════════════════════
int main(int argc, char** argv) {
    size_t numIters = DEFAULT_ITERS;
    if (argc > 1) numIters = static_cast<size_t>(atoi(argv[1]));

    // Generate test data
    fprintf(stderr, "Generating data...\n");
    std::vector<std::vector<std::vector<uint8_t>>> strCols(4);
    std::vector<std::vector<taper::VarcharSlice>> slices(4);
    for (size_t c = 0; c < 4; c++) {
        strCols[c].resize(NUM_OPS);
        slices[c].resize(NUM_OPS);
        for (size_t i = 0; i < NUM_OPS; i++) {
            auto s = "key_" + std::to_string(i) + "_c" + std::to_string(c);
            strCols[c][i].assign(s.begin(), s.end());
            slices[c][i].ptr = strCols[c][i].data();
            slices[c][i].len = strCols[c][i].size();
        }
    }

    // Pre-serialize for compare bench
    taper::SimpleArenaAllocator setupPool;
    std::vector<const uint8_t*> arenaBlocks(NUM_OPS);
    for (size_t i = 0; i < NUM_OPS; i++) {
        size_t totalSize = 0;
        for (size_t c = 0; c < 4; c++)
            totalSize += 1 + taper::ComputeRowLenSize(slices[c][i].len) + slices[c][i].len;
        uint8_t* block = setupPool.Allocate(static_cast<int64_t>(totalSize));
        uint8_t* wp = block;
        for (size_t c = 0; c < 4; c++)
            wp += taper::SerializeVarcharToBuffer(wp, slices[c][i].ptr, slices[c][i].len);
        arenaBlocks[i] = block;
    }

    // Hashes for probe bench
    std::mt19937_64 rng(42);
    std::vector<int64_t> hashes(NUM_OPS);
    for (size_t i = 0; i < NUM_OPS; i++) hashes[i] = static_cast<int64_t>(rng());
    size_t numChunks = 1;
    while (numChunks * 8 < NUM_OPS) numChunks *= 2;

    fprintf(stderr, "Done. Running %zu iters...\n\n", numIters);

    auto bench = [&](const char* name, auto fn) {
        volatile uint64_t w = fn(); (void)w; // warmup
        auto t0 = Clock::now();
        volatile uint64_t checksum = 0;
        for (size_t i = 0; i < numIters; i++) checksum = fn();
        auto t1 = Clock::now();
        double per_iter = elapsed_ms(t0, t1) / numIters;
        printf("%-25s  per_iter=%7.3f ms  checksum=%lu\n", name, per_iter, (unsigned long)checksum);
    };

    printf("=== C++ Individual Function Bench (%zu ops, %zu iters) ===\n", NUM_OPS, numIters);
    bench("1. arena_alloc_only", RunArenaAllocOnly);
    bench("2. new_row_only", RunNewRowOnly);
    bench("3. serialize_4str_only", [&]() { return RunSerialize4strOnly(slices); });
    bench("4. compare_4str_only", [&]() { return RunCompare4strOnly(arenaBlocks, slices); });
    bench("5. hashmap_probe_only", [&]() { return RunHashmapProbeOnly(hashes, numChunks); });
    bench("6. setrowptr_only", RunSetGetRowPtrOnly);

    return 0;
}

File: rust/src/bin/micro_functions_bench.rs
//! Individual function micro benchmarks — prove each function is equally fast in C++ and Rust.
//!
//! Usage: ./micro_functions_bench [iterations]
//! Default: 10 iterations, 1,000,000 operations each

use std::time::Instant;
use taper_hashmap::column_marshaller::{
    serialize_varchar_to_buffer, compute_row_len_size, compare_varchar_from_row,
    compute_varchar_serialized_size,
};
use taper_hashmap::row_container::{RowContainer, ColumnKind};
use taper_hashmap::taper_hashmap::TaperHashMap;
use taper_hashmap::chunk::SlotValue;

const NUM_OPS: usize = 1_000_000;
const DEFAULT_ITERS: usize = 10;

// ─── 1. arena_alloc_only ─────────────────────────────────────────────────
#[inline(never)]
fn run_arena_alloc_only() -> u64 {
    let key_sizes = vec![0usize; 4];
    let kinds = vec![ColumnKind::Varchar; 4];
    let mut rc = RowContainer::with_kinds(&key_sizes, &kinds, 8);
    let mut checksum: u64 = 0;
    for _ in 0..NUM_OPS {
        let p = rc.arena_alloc(50);
        checksum = checksum.wrapping_add(p as u64);
    }
    checksum
}

// ─── 2. new_row_only ─────────────────────────────────────────────────────
#[inline(never)]
fn run_new_row_only() -> u64 {
    let key_sizes = vec![0usize; 4];
    let kinds = vec![ColumnKind::Varchar; 4];
    let mut rc = RowContainer::with_kinds(&key_sizes, &kinds, 8);
    let mut checksum: u64 = 0;
    for _ in 0..NUM_OPS {
        let row = rc.new_row();
        checksum = checksum.wrapping_add(row as u64);
    }
    checksum
}

// ─── 3. serialize_4str_only ──────────────────────────────────────────────
#[inline(never)]
fn run_serialize_4str_only(str_cols: &[Vec<Vec<u8>>]) -> u64 {
    let key_sizes = vec![0usize; 4];
    let kinds = vec![ColumnKind::Varchar; 4];
    let mut rc = RowContainer::with_kinds(&key_sizes, &kinds, 8);
    let mut checksum: u64 = 0;
    for i in 0..NUM_OPS {
        let mut total_size = 0usize;
        for c in 0..4 {
            let s = &str_cols[c][i];
            total_size += 1 + compute_row_len_size(s.len()) as usize + s.len();
        }
        let block = rc.arena_alloc(total_size);
        let mut wp = block;
        for c in 0..4 {
            let s = &str_cols[c][i];
            let written = serialize_varchar_to_buffer(wp, s.as_slice());
            wp = unsafe { wp.add(written) };
        }
        checksum = checksum.wrapping_add(block as u64);
    }
    checksum
}

// ─── 4. compare_4str_only ────────────────────────────────────────────────
#[inline(never)]
fn run_compare_4str_only(arena_blocks: &[*const u8], str_cols: &[Vec<Vec<u8>>]) -> u64 {
    let mut match_count: u64 = 0;
    for i in 0..NUM_OPS {
        let mut pos = arena_blocks[i];
        let mut all_match = true;
        for c in 0..4 {
            let s = &str_cols[c][i];
            if !compare_varchar_from_row(pos, s.as_slice()) {
                all_match = false;
                break;
            }
            let entry_size = compute_varchar_serialized_size(pos);
            pos = unsafe { pos.add(entry_size) };
        }
        if all_match { match_count += 1; }
    }
    match_count
}

// ─── 5. hashmap_probe_only ───────────────────────────────────────────────
#[inline(never)]
fn run_hashmap_probe_only(hashes: &[u64], num_chunks: usize) -> u64 {
    let mut table = TaperHashMap::with_capacity(num_chunks);
    let mut sum: u64 = 0;
    table.emplace_batch_full(
        hashes,
        &|_: usize, _: &SlotValue| -> bool { true },
        &mut |_i: usize, sv: &mut SlotValue| { sv.bytes = [0x42; 6]; },
        &mut |_i: usize, _sv: &SlotValue, is_new: bool| { if !is_new { sum += 1; } },
    );
    table.len() as u64 + sum
}

// ─── 6. setrowptr_only (set_ptr / get_ptr) ──────────────────────────────
#[inline(never)]
fn run_set_get_ptr_only() -> u64 {
    let mut sv = SlotValue { bytes: [0u8; 6] };
    let mut checksum: u64 = 0;
    for i in 0..NUM_OPS {
        let ptr = (0x7FFF00000000u64 + (i as u64) * 41) as *const u8;
        sv.set_ptr(ptr);
        let got = sv.get_ptr();
        checksum = checksum.wrapping_add(got as u64);
    }
    checksum
}

// ═══════════════════════════════════════════════════════════════════
fn main() {
    let num_iters: usize = std::env::args()
        .nth(1)
        .and_then(|s| s.parse().ok())
        .unwrap_or(DEFAULT_ITERS);

    eprintln!("Generating data...");

    // String data
    let str_cols: Vec<Vec<Vec<u8>>> = (0..4)
        .map(|c| {
            (0..NUM_OPS)
                .map(|i| format!("key_{}_c{}", i, c).into_bytes())
                .collect()
        })
        .collect();

    // Pre-serialize for compare bench — rc must live as long as arena_blocks
    let key_sizes_cmp = vec![0usize; 4];
    let kinds_cmp = vec![ColumnKind::Varchar; 4];
    let mut rc_for_compare = RowContainer::with_kinds(&key_sizes_cmp, &kinds_cmp, 8);
    let arena_blocks: Vec<*const u8> = (0..NUM_OPS)
        .map(|i| {
            let mut total_size = 0usize;
            for c in 0..4 {
                let s = &str_cols[c][i];
                total_size += 1 + compute_row_len_size(s.len()) as usize + s.len();
            }
            let block = rc_for_compare.arena_alloc(total_size);
            let mut wp = block;
            for c in 0..4 {
                let s = &str_cols[c][i];
                let written = serialize_varchar_to_buffer(wp, s.as_slice());
                wp = unsafe { wp.add(written) };
            }
            block as *const u8
        })
        .collect();

    // Hashes for probe bench
    use rand_mt::Mt19937GenRand64;
    let mut rng = Mt19937GenRand64::new(42);
    let hashes: Vec<u64> = (0..NUM_OPS).map(|_| rng.next_u64()).collect();
    let mut num_chunks = 1usize;
    while num_chunks * 8 < NUM_OPS { num_chunks *= 2; }

    eprintln!("Done. Running {} iters...\n", num_iters);

    let bench = |name: &str, f: &dyn Fn() -> u64| {
        let _ = f(); // warmup
        let t0 = Instant::now();
        let mut checksum: u64 = 0;
        for _ in 0..num_iters { checksum = f(); }
        let per_iter = t0.elapsed().as_secs_f64() * 1000.0 / num_iters as f64;
        println!("{:<25}  per_iter={:7.3} ms  checksum={}", name, per_iter, checksum);
    };

    println!("=== Rust Individual Function Bench ({} ops, {} iters) ===", NUM_OPS, num_iters);
    bench("1. arena_alloc_only", &|| run_arena_alloc_only());
    bench("2. new_row_only", &|| run_new_row_only());
    bench("3. serialize_4str_only", &|| run_serialize_4str_only(&str_cols));
    bench("4. compare_4str_only", &|| run_compare_4str_only(&arena_blocks, &str_cols));
    bench("5. hashmap_probe_only", &|| run_hashmap_probe_only(&hashes, num_chunks));
    bench("6. setrowptr_only", &|| run_set_get_ptr_only());
}
